File: src/software/ai/hl/stp/tactic/move/move_tactic_test.cpp
#include "software/ai/hl/stp/tactic/move/move_tactic.h"

#include <gtest/gtest.h>

#include <utility>

#include "software/geom/algorithms/contains.h"
#include "software/simulated_tests/simulated_er_force_sim_play_test_fixture.h"
#include "software/simulated_tests/terminating_validation_functions/ball_kicked_validation.h"
#include "software/simulated_tests/terminating_validation_functions/robot_state_validation.h"
#include "software/simulated_tests/validation/validation_function.h"
#include "software/test_util/test_util.h"
#include "software/time/duration.h"
#include "software/world/world.h"

class MoveTacticTest : public SimulatedErForceSimPlayTestFixture
{
   protected:
    TbotsProto::FieldType field_type = TbotsProto::FieldType::DIV_B;
    Field field                      = Field::createField(field_type);
};

TEST_F(MoveTacticTest, test_move_across_field)
{
    Point initial_position = Point(-3, 1.5);
    Point destination      = Point(2.5, -1.1);
    BallState ball_state(Point(4.5, -3), Vector(0, 0));
    auto friendly_robots =
        TestUtil::createStationaryRobotStatesWithId({Point(-3, 2.5), initial_position});
    auto enemy_robots = TestUtil::createStationaryRobotStatesWithId(
        {Point(1, 0), Point(1, 2.5), Point(1, -2.5), field.enemyGoalCenter(),
         field.enemyDefenseArea().negXNegYCorner(),
         field.enemyDefenseArea().negXPosYCorner()});

    auto tactic = std::make_shared<MoveTactic>();
    tactic->updateControlParams(destination, Angle::zero(), 0, TbotsProto::MaxAllowedSpeedMode::PHYSICAL_LIMIT, TbotsProto::ObstacleAvoidanceMode::SAFE);
    setTactic(1, tactic);

    std::vector<ValidationFunction> terminating_validation_functions = {
        [destination, tactic](std::shared_ptr<World> world_ptr,
                              ValidationCoroutine::push_type& yield) {
            while (!tactic->done())
            {
                yield("Tactic not done");
            }
            robotAtPosition(1, world_ptr, destination, 0.05, yield);
            // Check that conditions hold for 1000 ticks
            unsigned num_ticks = 1000;
            for (unsigned i = 0; i < num_ticks; i++)
            {
                robotAtPosition(1, world_ptr, destination, 0.05, yield);
            }
        }};

    std::vector<ValidationFunction> non_terminating_validation_functions = {};

    runTest(field_type, ball_state, friendly_robots, enemy_robots,
            terminating_validation_functions, non_terminating_validation_functions,
            Duration::fromSeconds(10));
}

TEST_F(MoveTacticTest, test_autochip_move)
{
    Point initial_position = Point(-3, 1.5);
    Point destination      = Point(0, 1.5);
    BallState ball_state(Point(0, 1.5), Vector(0, 0));
    auto friendly_robots =
        TestUtil::createStationaryRobotStatesWithId({Point(-3, 2.5), initial_position});
    auto enemy_robots = TestUtil::createStationaryRobotStatesWithId(
        {Point(1, 0), Point(1, 2.5), Point(1, -2.5), field.enemyGoalCenter(),
         field.enemyDefenseArea().negXNegYCorner(),
         field.enemyDefenseArea().negXPosYCorner()});

    auto tactic = std::make_shared<MoveTactic>();
    tactic->updateControlParams(
        destination, Angle::zero(), 0,
        TbotsProto::DribblerMode::OFF,
        TbotsProto::BallCollisionType::ALLOW, {AutoChipOrKickMode::AUTOCHIP, 2.0},
        TbotsProto::MaxAllowedSpeedMode::COLLISIONS_ALLOWED,
        TbotsProto::ObstacleAvoidanceMode::SAFE, 0.0);
    setTactic(1, tactic);

    std::vector<ValidationFunction> terminating_validation_functions = {
        [destination, tactic](std::shared_ptr<World> world_ptr,
                              ValidationCoroutine::push_type& yield) {
            while (!tactic->done())
            {
                yield("Tactic not done");
            }
            robotAtPosition(1, world_ptr, destination, 0.05, yield);
            ballKicked(Angle::zero(), world_ptr, yield);
            // Check that conditions hold for 1000 ticks
            unsigned num_ticks = 1000;
            for (unsigned i = 0; i < num_ticks; i++)
            {
                robotAtPosition(1, world_ptr, destination, 0.05, yield);
            }
        }};

    std::vector<ValidationFunction> non_terminating_validation_functions = {};

    runTest(field_type, ball_state, friendly_robots, enemy_robots,
            terminating_validation_functions, non_terminating_validation_functions,
            Duration::fromSeconds(10));
}

TEST_F(MoveTacticTest, test_autokick_move)
{
    Point initial_position = Point(-1, -0.5);
    Point destination      = Point(-1, -1);
    BallState ball_state(Point(-1, -1), Vector(0, 0));
    auto friendly_robots = {RobotStateWithId{
        .id          = 0,
        .robot_state = RobotState(initial_position, Vector(0, 0), Angle::threeQuarter(),
                                  AngularVelocity::zero())}};
    auto enemy_robots    = TestUtil::createStationaryRobotStatesWithId(
        {Point(1, 0), Point(1, 2.5), Point(1, -2.5), field.enemyGoalCenter(),
         field.enemyDefenseArea().negXNegYCorner(),
         field.enemyDefenseArea().negXPosYCorner()});

    auto tactic = std::make_shared<MoveTactic>();
    tactic->updateControlParams(
        destination, Angle::threeQuarter(), 0, TbotsProto::DribblerMode::OFF,
        TbotsProto::BallCollisionType::ALLOW, {AutoChipOrKickMode::AUTOKICK, 3.0},
        TbotsProto::MaxAllowedSpeedMode::COLLISIONS_ALLOWED,
        TbotsProto::ObstacleAvoidanceMode::SAFE, 0.0);
    setTactic(0, tactic);

    std::vector<ValidationFunction> terminating_validation_functions = {
        [destination, tactic](std::shared_ptr<World> world_ptr,
                              ValidationCoroutine::push_type& yield) {
            while (!tactic->done())
            {
                yield("Tactic not done");
            }
            robotAtPosition(0, world_ptr, destination, 0.05, yield);
            ballKicked(Angle::threeQuarter(), world_ptr, yield);
            // Check that conditions hold for 1000 ticks
            unsigned num_ticks = 1000;
            for (unsigned i = 0; i < num_ticks; i++)
            {
                robotAtPosition(0, world_ptr, destination, 0.05, yield);
            }
        }};

    std::vector<ValidationFunction> non_terminating_validation_functions = {};

    runTest(field_type, ball_state, friendly_robots, enemy_robots,
            terminating_validation_functions, non_terminating_validation_functions,
            Duration::fromSeconds(10));
}

TEST_F(MoveTacticTest, test_spinning_move_clockwise)
{
    Point initial_position = Point(-4, 2);
    Point destination      = Point(4, 2);
    BallState ball_state(Point(1, 1), Vector(0, 0));
    auto friendly_robots = {RobotStateWithId{
        .id          = 0,
        .robot_state = RobotState(initial_position, Vector(0, 0), Angle::zero(),
                                  AngularVelocity::quarter())}};
    auto enemy_robots    = TestUtil::createStationaryRobotStatesWithId({Point(4, 0)});

    auto tactic = std::make_shared<MoveTactic>();
    tactic->updateControlParams(
        destination, Angle::zero(), 0, TbotsProto::DribblerMode::OFF,
        TbotsProto::BallCollisionType::ALLOW, {AutoChipOrKickMode::OFF, 0.0},
        TbotsProto::MaxAllowedSpeedMode::COLLISIONS_ALLOWED,
        TbotsProto::ObstacleAvoidanceMode::SAFE, 0.0);
    setTactic(0, tactic);

    std::vector<ValidationFunction> terminating_validation_functions = {
        [destination, tactic](std::shared_ptr<World> world_ptr,
                              ValidationCoroutine::push_type& yield) {
            robotAtAngularVelocity(0, world_ptr, AngularVelocity::fromDegrees(1 * 360),
                                   AngularVelocity::fromDegrees(50), yield);
            robotAtPosition(0, world_ptr, destination, 0.05, yield);
            robotAtOrientation(0, world_ptr, Angle::zero(), Angle::fromDegrees(5), yield);
            while (!tactic->done())
            {
                yield("Tactic is not done");
            }
            // Check that conditions hold for 1000 ticks
            unsigned num_ticks = 1000;
            for (unsigned i = 0; i < num_ticks; i++)
            {
                robotAtPosition(0, world_ptr, destination, 0.05, yield);
                robotAtOrientation(0, world_ptr, Angle::zero(), Angle::fromDegrees(5),
                                   yield);
            }
        }};

    std::vector<ValidationFunction> non_terminating_validation_functions = {};

    runTest(field_type, ball_state, friendly_robots, enemy_robots,
            terminating_validation_functions, non_terminating_validation_functions,
            Duration::fromSeconds(10));
}

TEST_F(MoveTacticTest, test_spinning_move_counter_clockwise)
{
    Point initial_position = Point(4, 2);
    Point destination      = Point(-4, 2);
    BallState ball_state(Point(1, 1), Vector(0, 0));
    auto friendly_robots = {RobotStateWithId{
        .id          = 0,
        .robot_state = RobotState(initial_position, Vector(0, 0), Angle::quarter(),
                                  AngularVelocity::zero())}};
    auto enemy_robots    = TestUtil::createStationaryRobotStatesWithId({Point(4, 0)});

    auto tactic = std::make_shared<MoveTactic>();
    tactic->updateControlParams(
        destination, Angle::half(), 0, TbotsProto::DribblerMode::OFF,
        TbotsProto::BallCollisionType::ALLOW, {AutoChipOrKickMode::OFF, 0.0},
        TbotsProto::MaxAllowedSpeedMode::PHYSICAL_LIMIT,
        TbotsProto::ObstacleAvoidanceMode::SAFE, -4.0);
    setTactic(0, tactic);

    std::vector<ValidationFunction> terminating_validation_functions = {
        [destination, tactic](std::shared_ptr<World> world_ptr,
                              ValidationCoroutine::push_type& yield) {
            robotAtAngularVelocity(0, world_ptr, AngularVelocity::fromDegrees(-4 * 360),
                                   AngularVelocity::fromDegrees(50), yield);
            robotAtPosition(0, world_ptr, destination, 0.05, yield);
            robotAtOrientation(0, world_ptr, Angle::half(), Angle::fromDegrees(5), yield);
            while (!tactic->done())
            {
                yield("Tactic is not done");
            }
            // Check that conditions hold for 1000 ticks
            unsigned num_ticks = 1000;
            for (unsigned i = 0; i < num_ticks; i++)
            {
                robotAtPosition(0, world_ptr, destination, 0.05, yield);
                robotAtOrientation(0, world_ptr, Angle::half(), Angle::fromDegrees(5),
                                   yield);
            }
        }};

    std::vector<ValidationFunction> non_terminating_validation_functions = {};

    runTest(field_type, ball_state, friendly_robots, enemy_robots,
            terminating_validation_functions, non_terminating_validation_functions,
            Duration::fromSeconds(10));
}

File: src/software/ai/hl/stp/tactic/move/move_tactic_test.py
import pytest

import software.python_bindings as tbots_cpp
from software.simulated_tests.pytest_validations.robot_at_position import (
    RobotEventuallyAtPosition,
)
from software.simulated_tests.pytest_validations.ball_kicked_in_direction import (
    BallEventuallyKickedInDirection,
)
from software.simulated_tests.pytest_validations.robot_at_orientation import (
    RobotEventuallyAtOrientation,
)
from software.simulated_tests.pytest_validations.ball_is_off_ground import (
    BallIsEventuallyOffGround,
)
from software.simulated_tests.pytest_validations.robot_at_angular_velocity import (
    RobotEventuallyAtAngularVelocity,
)
from software.simulated_tests.simulated_test_fixture import (
    pytest_main,
)
from proto.message_translation.tbots_protobuf import create_world_state
from proto.import_all_protos import *


def test_move_across_field(simulated_test_runner):
    initial_position = tbots_cpp.Point(-3, 1.5)
    destination = tbots_cpp.Point(2.5, -1.1)
    field = tbots_cpp.Field.createSSLDivisionBField()

    def setup(*args):
        simulated_test_runner.set_world_state(
            create_world_state(
                blue_robot_locations=[
                    tbots_cpp.Point(-3, 2.5),
                    initial_position,
                ],
                yellow_robot_locations=[
                    tbots_cpp.Point(1, 0),
                    tbots_cpp.Point(1, 2.5),
                    tbots_cpp.Point(1, -2.5),
                    field.enemyGoalCenter(),
                    field.enemyDefenseArea().negXNegYCorner(),
                    field.enemyDefenseArea().negXPosYCorner(),
                ],
                ball_location=tbots_cpp.Point(4.5, -3),
                ball_velocity=tbots_cpp.Vector(0, 0),
            ),
        )

        simulated_test_runner.set_tactics(
            blue_tactics={
                1: MoveTactic(
                    destination=tbots_cpp.createPointProto(destination),
                    final_orientation=tbots_cpp.createAngleProto(
                        tbots_cpp.Angle.zero()
                    ),
                )
            }
        )

    eventually_validation_sequence_set = [
        [
            # TODO: should also validate that the robot stays at destination for 1000 ticks after
            RobotEventuallyAtPosition(1, destination),
        ]
    ]

    simulated_test_runner.run_test(
        setup=setup,
        inv_eventually_validation_sequence_set=eventually_validation_sequence_set,
        ag_eventually_validation_sequence_set=eventually_validation_sequence_set,
    )


def test_autochip_move(simulated_test_runner):
    initial_position = tbots_cpp.Point(-3, 1.5)
    destination = tbots_cpp.Point(0, 1.5)
    field = tbots_cpp.Field.createSSLDivisionBField()

    def setup(*args):
        simulated_test_runner.set_world_state(
            create_world_state(
                blue_robot_locations=[
                    tbots_cpp.Point(-3, 2.5),
                    initial_position,
                ],
                yellow_robot_locations=[
                    tbots_cpp.Point(1, 0),
                    tbots_cpp.Point(1, 2.5),
                    tbots_cpp.Point(1, -2.5),
                    field.enemyGoalCenter(),
                    field.enemyDefenseArea().negXNegYCorner(),
                    field.enemyDefenseArea().negXPosYCorner(),
                ],
                ball_location=destination,
                ball_velocity=tbots_cpp.Vector(0, 0),
            ),
        )

        simulated_test_runner.set_tactics(
            blue_tactics={
                1: MoveTactic(
                    destination=tbots_cpp.createPointProto(destination),
                    final_orientation=tbots_cpp.createAngleProto(
                        tbots_cpp.Angle.zero()
                    ),
                    dribbler_mode=DribblerMode.OFF,
                    ball_collision_type=BallCollisionType.ALLOW,
                    auto_chip_or_kick=AutoChipOrKick(autochip_distance_meters=2.0),
                    max_allowed_speed_mode=MaxAllowedSpeedMode.COLLISIONS_ALLOWED,
                    obstacle_avoidance_mode=ObstacleAvoidanceMode.SAFE,
                )
            }
        )

    eventually_validation_sequence_set = [
        [
            # TODO (#2558): should also validate that the robot stays at destination for 1000 ticks after
            RobotEventuallyAtPosition(1, destination),
            BallEventuallyKickedInDirection(tbots_cpp.Angle.zero()),
            BallIsEventuallyOffGround(),
        ]
    ]

    simulated_test_runner.run_test(
        setup=setup,
        inv_eventually_validation_sequence_set=eventually_validation_sequence_set,
        ag_eventually_validation_sequence_set=eventually_validation_sequence_set,
        test_timeout_s=10,
    )


def test_autokick_move(simulated_test_runner):
    initial_position = tbots_cpp.Point(-1, -0.5)
    destination = tbots_cpp.Point(-1, -1)
    field = tbots_cpp.Field.createSSLDivisionBField()

    def setup(*args):
        simulated_test_runner.set_world_state(
            create_world_state(
                blue_robot_locations=[initial_position],
                blue_robot_orientations=[tbots_cpp.Angle.threeQuarter()],
                yellow_robot_locations=[
                    tbots_cpp.Point(1, 0),
                    tbots_cpp.Point(1, 2.5),
                    tbots_cpp.Point(1, -2.5),
                    field.enemyGoalCenter(),
                    field.enemyDefenseArea().negXNegYCorner(),
                    field.enemyDefenseArea().negXPosYCorner(),
                ],
                ball_location=destination,
                ball_velocity=tbots_cpp.Vector(0, 0),
            ),
        )

        simulated_test_runner.set_tactics(
            blue_tactics={
                0: MoveTactic(
                    destination=tbots_cpp.createPointProto(destination),
                    final_orientation=tbots_cpp.createAngleProto(
                        tbots_cpp.Angle.threeQuarter()
                    ),
                    dribbler_mode=DribblerMode.OFF,
                    ball_collision_type=BallCollisionType.ALLOW,
                    auto_chip_or_kick=AutoChipOrKick(autokick_speed_m_per_s=3.0),
                    max_allowed_speed_mode=MaxAllowedSpeedMode.COLLISIONS_ALLOWED,
                    obstacle_avoidance_mode=ObstacleAvoidanceMode.SAFE,
                )
            }
        )

    eventually_validation_sequence_set = [
        [
            # TODO (#2558): should also validate that the robot stays at destination for 1000 ticks after
            RobotEventuallyAtPosition(0, destination),
            BallEventuallyKickedInDirection(tbots_cpp.Angle.threeQuarter()),
        ]
    ]

    simulated_test_runner.run_test(
        setup=setup,
        inv_eventually_validation_sequence_set=eventually_validation_sequence_set,
        ag_eventually_validation_sequence_set=eventually_validation_sequence_set,
        test_timeout_s=5,
    )


@pytest.mark.parametrize(
    "orientation, initial_position, destination, angular_velocity",
    [
        # Robot facing right, should rotate counter clockwise to face up
        (
            tbots_cpp.Angle.quarter(),
            tbots_cpp.Point(2, 0),
            tbots_cpp.Point(-2, 0),
            tbots_cpp.Angle.fromDegrees(360),
        ),
        # Robot facing right, should rotate clockwise to face down
        (
            tbots_cpp.Angle.threeQuarter(),
            tbots_cpp.Point(2, 0),
            tbots_cpp.Point(-2, 0),
            tbots_cpp.Angle.fromDegrees(-360),
        ),
        # Robot facing right, should rotate counter clockwise to face up
        (
            tbots_cpp.Angle.quarter(),
            tbots_cpp.Point(0, 0),
            tbots_cpp.Point(0, 0),
            tbots_cpp.Angle.fromDegrees(360),
        ),
        # Robot facing right, should rotate clockwise to face down
        (
            tbots_cpp.Angle.threeQuarter(),
            tbots_cpp.Point(0, 0),
            tbots_cpp.Point(0, 0),
            tbots_cpp.Angle.fromDegrees(-360),
        ),
        # Robot facing right, should rotate counter clockwise to face left, slightly up
        (
            tbots_cpp.Angle.fromDegrees(175),
            tbots_cpp.Point(3, -1),
            tbots_cpp.Point(0, 1),
            tbots_cpp.Angle.fromDegrees(360),
        ),
        # Robot facing right, should rotate clockwise to face down
        (
            tbots_cpp.Angle.fromDegrees(185),
            tbots_cpp.Point(3, -1),
            tbots_cpp.Point(0, 1),
            tbots_cpp.Angle.fromDegrees(-360),
        ),
    ],
)
def test_spinning_move(
    orientation, initial_position, destination, angular_velocity, simulated_test_runner
):
    def setup(*args):
        simulated_test_runner.set_world_state(
            create_world_state(
                blue_robot_locations=[initial_position],
                yellow_robot_locations=[tbots_cpp.Point(4, 0)],
                ball_location=tbots_cpp.Point(1, 1),
                ball_velocity=tbots_cpp.Vector(0, 0),
            ),
        )

        simulated_test_runner.set_tactics(
            blue_tactics={
                0: MoveTactic(
                    destination=tbots_cpp.createPointProto(destination),
                    final_orientation=tbots_cpp.createAngleProto(orientation),
                    dribbler_mode=DribblerMode.OFF,
                    ball_collision_type=BallCollisionType.ALLOW,
                    auto_chip_or_kick=AutoChipOrKick(),
                    max_allowed_speed_mode=MaxAllowedSpeedMode.PHYSICAL_LIMIT,
                    obstacle_avoidance_mode=ObstacleAvoidanceMode.SAFE,
                )
            }
        )

    eventually_validation_sequence_set = [
        [
            # high threshold just to check direction of angular velocity
            RobotEventuallyAtAngularVelocity(0, angular_velocity, 4),
            RobotEventuallyAtPosition(0, destination),
            RobotEventuallyAtOrientation(0, orientation),
            # TODO: validate position and orientation for 1000 ticks
        ]
    ]

    simulated_test_runner.run_test(
        setup=setup,
        inv_eventually_validation_sequence_set=eventually_validation_sequence_set,
        ag_eventually_validation_sequence_set=eventually_validation_sequence_set,
    )


if __name__ == "__main__":
    pytest_main(__file__)
